man/ssocksd.1
.\"   ssocksd.1
.\"
.\"   Man page for nsocks
.\"
.TH ssocksd 1 "3 Apr 2011" "ssocks" "ssocksd manual"
.SH NAME
ssocksd \- Get information about ssocksd
.SH SYNOPSIS
.B ssocksd
.SH DESCRIPTION
.B ssocksd
Socks5 server, implements RFC 1928 (SOCKS V5) and RFC 1929 
(Authentication for SOCKS V5)

Features:

	- Configuration file

	- Authentication file

	- Generate a connection log file

	- Daemon mode

	- Server support bind mode



.SH OPTIONS
.IP "--verbose"
Increase the verbose level, can be add twice for debug mode
.IP "--daemon"
Daemon mode, no ouput, and background
.IP "--post port"
Listening port
.IP "--auth passwdfile"
Password file see ssocksd.auth (5) for format
.IP "--log logfile"
If is set a user connection log file is write
.IP "--config configfile"
If is set the configuration is load see ssocksd.conf (5) for format
	
.SH "EXAMPLES"
Listen on port 1080, no authentification, no log file

	$ ssocksd

Listen on port 8080, and ask for authenfication with credential of file /etc/ssocksd.auth

	$ ssocksd --auth /etc/ssocksd.auth --port 8080

Load configuration file and execute the server in daemon mode 
(if -d is after -f is bypass the option daemon of the config file)

	$ ssocksd -f /etc/ssocksd.conf -d

.SH "SEE ALSO"
.BR socksd (1), nsocks (1), ssocks-keygen (1), ssocksd.auth (5), ssocksd.conf (5), rssocks (1)


NEWS
	2011/04/18
Release v0.0.10:
	- Fix, non blocking bug on ssocksd 
	- True support of request address type IPv4
	- Add socks4 support, fast implementation but should works ( if auth is 
	enable socks4 is disable )
	- Add to rssocks a parameter --ncon ( -n ) to set the number of 
	back connection	you want ( default 25 )
	- Set non blocking connect on rssocks
	
	
-------------------------------------------------------------------------------
	2011/04/17
Upload video, how to use rssocks and rcsocks in pentest mode
http://www.vimeo.com/22515255

-------------------------------------------------------------------------------
	2011/04/16
Release v0.0.9:
	- New code, now the socks code is in libsocks
	- No more thread (now when it connect it does this with non blocking method)
	- Better stability
	- Log format change
	- Add rcsocks a client for reverse socks 
	- Add rssocks a server for reverse socks to bypass firewall restriction
	- No SSL in this version is under developpement
-------------------------------------------------------------------------------
	2011/04/08
Release v0.0.8-dev-ssl:
	- First dev version with openssl support
	- To use it you need to generate server private key and certificate for this
	I've add a tool, ssocks-keygen is bash script ( man ssocks-keygen )

	Example of use:
	  test@remoteserver$ ssocksd -s -c keys/server.crt -k keys/server.key \
			-a ssocksd.auth
	
	  you@laptop$ ssocks --ca src/keys/ca.crt -s remoteserver:1080 \
			-u admin -p abcde -l 1088 

	Now set your browser to use localhost 1088 with protocole socks5.
	All you'r request pass trought the localsocks
	that handles connect to remote socks5 server establish a ssl tunnel, 
	check the certificate, authenticate you with admin:abcde, and transmit data.

-------------------------------------------------------------------------------
	2011/04/07
Release v0.0.7:
	- First stable version
	- Use a thread when the server do the request to avoid blocking
	- Configuration file
	- Authentication file
	- Generate a connection log file
	- Daemon mode
	- Server support bind mode


src/ssocksd.c
/*
 *      ssocksd.c
 *      
 *      Created on: 2011-03-30
 *      Author:     Hugo Caron
 *      Email:      <h.caron@codsec.com>
 * 
 * Copyright (C) 2011 by Hugo Caron
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#include <libsocks/net-util.h>
#include <libsocks/bor-util.h>
#include <libsocks/output-util.h>
#include <libsocks/client.h>

#include <libsocks/socks-common.h>
#include <libsocks/socks5-server.h>
#include <libsocks/log-util.h>

#include "auth-util.h"

#include "configd-util.h"
#include "auth-util.h"

#include <getopt.h>
#include <unistd.h>
#include <config.h>

#define DEFAULT_PORT 1080
#define PID_FILE "/var/run/ssocksd.pid"

int boucle_princ = 1;
void capte_fin (int sig){
    printf ("serveur: signal %d caught\n", sig);
    boucle_princ = 0;
}

/* TODO: Add --pid-file option to server
 */
void usage(char *name){
	printf("ssockd - Server Socks5 v%s\n", PACKAGE_VERSION);

	printf("Usage:\n");
	printf("\t%s --port 8080\n", name);
	printf("\t%s -p 8080 -a ssocksd.auth -d\n", name);
	printf("\t%s -vv\n", name);
	printf("\n");
	printf("Options:\n");
#ifdef HAVE_LIBSSL
	printf("\t--ssl      enable secure socks5 protocol\n");
	printf("\t--cert {file.crt}   set server certificate\n");
	printf("\t--key  {file.pem}    set server private key\n");
#endif
	printf("\t--daemon   daemon mode (background)\n");
	printf("\t--verbose  increase verbose level\n\n");
	printf("\t--port {port}  listening port (default 1080)\n");
	printf("\t--file {file}  see man 5 ssocksd.conf\n");
	printf("\t--auth {file}  see man 5 ssocksd.auth\n");
	printf("\t--log {file}   if set connections are log in this file\n");
	printf("\n");
	printf("Bug report %s\n", PACKAGE_BUGREPORT);
}

void parseArg(int argc, char *argv[]){
	int c;
	
	globalArgsServer.fileauth[0] = 0;
	globalArgsServer.filelog[0] = 0;
	globalArgsServer.fileconfig[0] = 0;
	globalArgsServer.port = DEFAULT_PORT;
	globalArgsServer.verbosity = 0;
	globalArgsServer.guest = 1;

#ifdef HAVE_LIBSSL
	globalArgsServer.ssl = 0;
	globalArgsServer.filecert[0] = 0;
	globalArgsServer.filekey[0] = 0;
#endif

	
	while (1){
		static struct option long_options[] = {
			{"help",    no_argument,       0, 'h'},
			{"verbose", no_argument,       0, 'v'},
			{"daemon",  no_argument,       0, 'd'},
#ifdef HAVE_LIBSSL
			{"ssl",  no_argument,       0, 's'},
			{"cert", required_argument, 0, 'c'},
			{"key",  required_argument, 0, 'k'},
#endif
			{"guest",no_argument,       0, 'g'},
			{"port", required_argument, 0, 'p'},
			{"file", required_argument, 0, 'f'},
			{"auth", required_argument, 0, 'a'},
			{"log",  required_argument, 0, 'l'},
			{0, 0, 0, 0}
		};
		
		/* getopt_long stores the option index here. */
		int option_index = 0;

		c = getopt_long (argc, argv, "h?vsgdk:c:f:a:p:l:",
					long_options, &option_index);

		/* Detect the end of the options. */
		if (c == -1)
			break;
		/* TODO: Dangerous strcpy here fix this!!! */
		switch (c)	{
			case 0:
				/* If this option set a flag, do nothing else now. */
				if (long_options[option_index].flag != 0)
					break;
				printf ("option %s", long_options[option_index].name);
				if (optarg)
					printf (" with arg %s", optarg);
				printf ("\n");
				break;

			case 'v':
				globalArgsServer.verbosity++;
				verbosity++;
				break;

			case 'd':
				globalArgsServer.daemon = 1;
				break;

#ifdef HAVE_LIBSSL
			case 's':
				globalArgsServer.ssl = 1;
				break;
			case 'c':
				strcpy(globalArgsServer.filecert, optarg);
				break;
			case 'k':
				strcpy(globalArgsServer.filekey, optarg);
				break;
#endif
			case 'g':
				globalArgsServer.guest = 1;
				break;
				
			case 'p':
				globalArgsServer.port = atoi(optarg);
				break;

			case 'a':
				strcpy(globalArgsServer.fileauth, optarg);
				break;
				
			case 'l':
				strcpy(globalArgsServer.filelog, optarg);
				break;
				
			case 'f':
				strcpy(globalArgsServer.fileconfig, optarg);
				if ( loadConfigFile(optarg, &globalArgsServer) < 0 ){
					ERROR(L_NOTICE, "config: config file error");
					ERROR(L_NOTICE, "server: can't start configuration error");
					exit(1);
				}
				break;

			case '?':
				/* getopt_long already printed an error message. */
				usage(argv[0]);
				exit(1);
				break;
				
			case 'h':
				usage(argv[0]);
				exit(1);
				break;
				
			default:
				abort ();
		}
	}
	

	if ( globalArgsServer.filelog[0] != 0 ){
		open_log(globalArgsServer.filelog);
	}
	if ( globalArgsServer.fileauth[0] != 0 ){
		globalArgsServer.guest = 0;
		if ( (c = load_auth_file(globalArgsServer.fileauth)) <= 0 ){
			ERROR(L_NOTICE, "auth: no username load");
			ERROR(L_NOTICE, "server: can't start configuration error");
			exit(1);
		}else{
			TRACE(L_NOTICE, "auth: %d usernames load", c);
		}
	}else{
		TRACE(L_NOTICE, "warning: no authentication enable");
	}

#ifdef HAVE_LIBSSL
	if (globalArgsServer.ssl == 1){
		SSL_load_error_strings();  /* readable error messages */
		SSL_library_init();        /* initialize library */
		if ( globalArgsServer.filecert[0] == 0 ){
			ERROR(L_NOTICE, "server: need a certificate file to use ssl");
		}
		if ( globalArgsServer.filekey[0] == 0 ){
			ERROR(L_NOTICE, "server: need a private key file to use ssl");
		}
		if ( ssl_init_server(globalArgsServer.filecert,
				globalArgsServer.filekey, SSL_FILETYPE_PEM) != 0){
			ERROR(L_NOTICE, "server: ssl configuration error");
			exit(1);
		}
	}
#endif

	verbosity = globalArgsServer.verbosity;
}

void capte_usr1(){
	TRACE(L_DEBUG, "server: catch USR1 signal ...");
}
void capte_sigpipe(){
	TRACE(L_DEBUG, "server: catch SIGPIPE signal ...");
}
void server(int port){
    int soc_ec = -1, maxfd, res, nc;  
    s_client tc[MAXCLI];
    fd_set set_read;
    fd_set set_write;
    struct sockaddr_in addrS;
    char methods[2];
    
    s_socks_conf conf;
    s_socks_server_config config;
    conf.config.srv = &config;

    char versions[2] = { 0x05, 0x04 };
    config.allowed_version = versions;
    config.n_allowed_version = 2;

    if ( globalArgsServer.fileauth[0] != 0 ){
    	methods[0] = 0x02;
    	--config.n_allowed_version; /* Disable socks4 don't support auth */
    }else{
    	methods[0] = 0x00;
    }
    



    config.allowed_method = methods;
    config.n_allowed_method = 1;
    config.check_auth = check_auth;
    
    /* Init client tab */
    for (nc = 0; nc < MAXCLI; nc++) 
		init_client (&tc[nc], nc, M_SERVER, &conf);
    
    soc_ec = new_listen_socket (port, 0, &addrS);
    if (soc_ec < 0) goto fin_serveur;
    
	
	if ( globalArgsServer.daemon == 1 ){
		TRACE(L_NOTICE, "server: mode daemon ...");
		if ( daemon(0, 0) != 0 ){
			perror("daemon");
			exit(1);
		}
		writePID(PID_FILE);
	}    
    
    bor_signal (SIGINT, capte_fin, SA_RESTART);

    /* Need in daemon to remove the PID file properly */
    bor_signal (SIGTERM, capte_fin, SA_RESTART);
    bor_signal (SIGPIPE, capte_sigpipe, SA_RESTART);
    /* TODO: Find a better way to exit the select and recall the init_select
     * SIGUSR1 is send by a thread to unblock the select */
    bor_signal (SIGUSR1, capte_usr1, SA_RESTART);

    while (boucle_princ) {
        init_select_server (soc_ec, tc, &maxfd, &set_read, &set_write);
        
        res = select (maxfd+1, &set_read, &set_write, NULL, NULL);

        if (res > 0) { /* Search eligible sockets */
        
            if (FD_ISSET (soc_ec, &set_read))
                new_connection (soc_ec, tc);
            
            for (nc = 0; nc < MAXCLI; nc++){
				dispatch_server(&tc[nc], &set_read, &set_write);
			}
                
        } else if ( res == 0){
 
        }else if (res < 0) { 
            if (errno == EINTR) ; /* Received signal, it does nothing */
            else { perror ("select"); goto fin_serveur; }
        }
    }   

fin_serveur: 
#ifdef HAVE_LIBSSL
	if (globalArgsServer.ssl == 1)
		ssl_cleaning();
#endif
    TRACE(L_NOTICE, "server: closing sockets ...");
    close_log();
    for (nc = 0; nc < MAXCLI; nc++) disconnection(&tc[nc]);
    if (soc_ec != -1) close (soc_ec);
    if ( globalArgsServer.daemon == 1 )	removePID(PID_FILE);
}

int main (int argc, char *argv[]){
	parseArg(argc, argv);
	server(globalArgsServer.port);
    exit (0);
}
